mitiq/cdr/data_regression.py

```diff
@@ -0,0 +1,294 @@
+    on the training data.
+    Args:
+        X_data: array of noisy observables at various noise levels.
+        y_data: array of exact observables.
+        intercept: whether or not to include and intercept in the fitting
+                   function.
+    """
+    noise_levels = len(X_data[0][:])
+    X_data = X_data.T
+    if intercept:
+        params_0 = np.zeros((noise_levels + 1))
+    else:
+        params_0 = np.zeros((noise_levels))
+
+    # Here need to use some package that does the regression.
+    popt, pcov = curve_fit(
+        lambda x, *params: wrapper_fit_func(
+            x, noise_levels, params, intercept
+        ),
+        X_data,
+        y_data,
+        p0=params_0,
+    )
+    return popt
+
+
+def f_intercept(X_data: np.ndarray, params: List) -> float:
+    """ Function used to map noisy to exact expectation values. Fitted using
+    data from the training circuits.
+    Args:
+        X_data: array of noisy observables are various noise levels.
+        *params: parameters of function.
+    """
+    return sum(a * x for a, x in zip(params, X_data)) + params[-1]
+
+
+def f(X_data: np.ndarray, params: np.ndarray) -> float:
+    """ Function used to map noisy to exact expectation values. Fitted using
+    data from the training circuits. No intercept.
+    Args:
+        X_data: array of noisy observables are various noise levels.
+        *params: parameters of function.
+    """
+    return sum(a * x for a, x in zip(params, X_data))
+
+
+def wrapper_fit_func(x, noise_levels, *args, intercept: bool = True):
+    if intercept:
+        a = list(args[0][: noise_levels + 1])
+        return f_intercept(x, a)
+    else:
+        a = list(args[0][:noise_levels])
+        return f(x, a)
+
+
+# TODO: Discuss if this format for converting shots makes sense in general.
+def dictionary_to_state_counts(counts: Dict, Q: int) -> Dict:
+    """Expresses the result of the simulation in the form of a dictionary
+    whose values are the modulus squared of the components of the final state.
+    The return probabilities are normalized by the number of counts.
+    Args:
+        counts: Dictionary of counts with binary keys identifying the state.
+        Q: Number of qubits in the system.
+    Returns:
+        state: Dictionary whose keys are the base elements of the Q qubit
+        system and whose values are the modulus of the corresponding squared
+        amplitudes.
+    """
+    basis = {i: bin(i) for i in range(2 ** Q)}
+    counts_order = np.array([i for i in range(2 ** Q)])
+    for i in range(len(basis)):
+        key = list(basis.values())[i]
+        if key not in counts:
+            counts[key] = 0
+        for j in range(len(counts)):
+            if list(basis.values())[i] == list(counts.keys())[j]:
+                counts_order[i] = list(counts.values())[j]
+    state = {
+        list(basis.values())[i]: (counts_order[i] / sum(counts_order))
+        for i in range(2 ** Q)
+    }
+    return state
```

Comment:
Here we could use `collecions.Counter`, which should do exactly what we need:
```
>>> from collections import Counter
>>> z = ['blue', 'red', 'blue', 'yellow', 'blue', 'red']
>>> Counter(z)
Counter({'blue': 3, 'red': 2, 'yellow': 1})
```

Reply:
Actually I think you just want to add missing keys and normalize the values. So this should work:

```python
def dictionary_to_probabilities(
    counts: Dict[str, int], nqubits: int,
) -> Dict[str, float]:
    """Expresses the result of the simulation in the form of a dictionary
    whose values are the modulus squared of the components of the final state.
    The return probabilities are normalized by the number of counts.
    Args:
        counts: Dictionary of counts with binary keys identifying the state.
        nqubits: Number of qubits in the system.
    Returns:
        state: Dictionary whose keys are the base elements of the nqubits qubit
        system and whose values are the modulus of the corresponding squared
        amplitudes.
    """
    total_counts = sum(counts.values())
    # Initialize probabilities to 0.0
    state = {bin(j): 0.0 for j in range(2 ** nqubits)}
    # Update with normalized counts
    for key, value in counts.items():
        state[key] = value / total_counts
    return state
```

Reply:
By the way, can't we just return the array of probabilities instead of a dictionary?
Do we use the keys?

Reply:
hey, yes I think your right, we can just return the probabilities.